# Workflow/pipeline details for the Culex biting RNA-seq analysis

## Upstream 
Information about sequence reads (raw read count, read alignment rate, etc.) for the data can be found [here](https://docs.google.com/spreadsheets/d/1y15EVJ7VUNeKWtLNaBUMJ1zZaR_LLv7YgWeYWtGrIpI/edit?usp=sharing)

### Data Accession
Data was generated by collaborator...

Add accession notes here

### Preprocessing and Quality Control

Trimmomatic (version 0.39) was used to trim sequence reads based on quality ([script](https://github.com/srmarzec/Culex_Biting_RNAseq/blob/main/Upstream/trim.sh))

FastQC (v0.11.9) was used for quality control visualization ([script](https://github.com/srmarzec/Culex_Biting_RNAseq/blob/main/Upstream/fastqc.sh))

Preliminary trimming and fastqc showed a poor "per sequence base content" for the first ~15 bases. We decided to use HEADCROP flag to remove the first 15 bases. All other flags (TRAILING, SLIDINGWINDOW, and MINLEN) are rather general/default for basic quality of bases and did not result in much difference of trimming.

### Mapping

### Generating count matrix with HTSeq (htseq-count)

## Downstream

